### PR TITLE
vaccination schedule

### LINKED ISSUE #226: Move to pre-defined vaccination schedule
**Idea:** 
Instead of dynamically calculating how many people in each age group should be vaccinated at a given tilmestep, precalculate all this ahead of the simulation. This will
- make the forward simulation more consistent with what we want to do for fitting (where we have exact numbers of individuals in each age group who received a vaccine dose on each day)
- allow more flexibility in the prioritisation of groups e.g. will allow a fraction of young people (representing HCWs and people with underlying conditions) in an early priority group while the rest gets vaccinated later
- easier discounting of second doses from total available doses on a given day

**Schematic:**
IMG_2421

For each age group i (shown as columns in the schematic above) define 
- `p[i,j]` as the proportion of individuals in age group i who are in priority group j (`j = 0, ...., n_priority_groups`). Note that priority group 0 in the schematic above is used to represent reluctant people who will never get vaccinated (--> this means it won't be easy to model time varying uptake). 
- `n[i,j] = p[i,j] * population[i]` as the number of individuals in age group i and priority group j.

In the following we consider `p[i>1, j]` and `n[i>1, j]` i.e. we remove the reluctant people from p and n.

**Proposed algorithm to generate pre-defined schedule:**
```R 

################################################################################
## parameters
################################################################################

daily_doses <- rep(20000, 365) # a vector of number of doses to give each day
region <- "london"

n_age_groups <- 19
n_priority_groups <- 12 

################################################################################
## calculate proportion to vaccinate by age and priority group: p
################################################################################

p <- matrix(0, n_age_groups, n_priority_groups)

### assuming uptake is 
# 95% in CHR
# 85% in and CHW
# 95% in 80+
# 85% in 50-80
# 75% in <50
uptake_by_age <- c(rep(0, 3), # no vaccination in <15
                   (2/5) * 0.75, # no vaccination in 15-17yo
                   rep(0.75, 6),
                   rep(0.85, 6),
                   0.95, 
                   0.85, 
                   0.95)

## https://www.gov.uk/government/publications/priority-groups-for-coronavirus-covid-19-vaccination-advice-from-the-jcvi-30-december-2020/joint-committee-on-vaccination-and-immunisation-advice-on-priority-groups-for-covid-19-vaccination-30-december-2020#fnref:3
## Assuming independance between job (e.g. HCW) and clinical condition
## But assuming one is either counted as "clinically extremely vulnerable" or "with underlying health conditions" but not both

## JCVI group 1: residents in a care home for older adults and their carers
j <- 1
i <- 18:19
p[i, j] <- uptake_by_age[i]
## JCVI group 2: all those 80 years of age and over and frontline health and social care workers
j <- 2
i <- 17
prop_hcw_by_age <- c(rep(0, 4), 
                     rep(0.1, 10),
                     rep(0, 5)) # numbers made up for now
p[i, j] <- uptake_by_age[i]
p[1:(i-1), j] <- uptake_by_age[1:(i-1)] * prop_hcw_by_age[1:(i-1)]
## JCVI group 3: all those 75 years of age and over
j <- 3
i <- 16
p[i, j] <- uptake_by_age[i] * (1 - prop_hcw_by_age[i])
## JCVI group 4: all those 70 years of age and over and clinically extremely vulnerable individuals
j <- 4
i <- 15
prop_very_vulnerable_by_age <- c(rep(0, 4), 
                     rep(0.05, 5),
                     rep(0.1, 5),
                     rep(0.15, 5)) # numbers made up for now
p[i, j] <- uptake_by_age[i] * (1 - prop_hcw_by_age[i])
p[1:(i-1), j] <- uptake_by_age[1:(i-1)] * (1 - prop_hcw_by_age[1:(i-1)]) * prop_very_vulnerable_by_age[1:(i-1)]
## JCVI group 5: all those 65 years of age and over
j <- 5
i <- 14
p[i, j] <- uptake_by_age[i] * (1 - prop_hcw_by_age[i]) * (1 - prop_very_vulnerable_by_age[i])
## JCVI group 6: all individuals aged 16 years to 64 years with underlying health conditions which put them at higher risk of serious disease and mortality
j <- 6
i <- 4:13 # starting at 15 not 16 so will need adjusting proportion for this
prop_underlying_condition_by_age <- c(rep(0, 4), 
                                      rep(0.05, 5),
                                      rep(0.1, 5),
                                      rep(0.15, 5)) # numbers made up for now
p[i, j] <- uptake_by_age[i] * (1 - prop_hcw_by_age[i]) * prop_underlying_condition_by_age[i]
## JCVI group 7: all those 60 years of age and over
j <- 7
i <- 13
p[i, j] <- uptake_by_age[i] * (1 - prop_hcw_by_age[i]) * (1 - prop_very_vulnerable_by_age[i] - prop_underlying_condition_by_age[i])
## JCVI group 8: all those 55 years of age and over
j <- 8
i <- 12
p[i, j] <- uptake_by_age[i] * (1 - prop_hcw_by_age[i]) * (1 - prop_very_vulnerable_by_age[i] - prop_underlying_condition_by_age[i])
## JCVI group 9: all those 50 years of age and over
j <- 9
i <- 11
p[i, j] <- uptake_by_age[i] * (1 - prop_hcw_by_age[i]) * (1 - prop_very_vulnerable_by_age[i] - prop_underlying_condition_by_age[i])

## JCVI second phase
## JCVI group 10: all those 40-49 years of age and over
j <- 10
i <- 9:10
p[i, j] <- uptake_by_age[i] * (1 - prop_hcw_by_age[i]) * (1 - prop_very_vulnerable_by_age[i] - prop_underlying_condition_by_age[i])
## JCVI group 11: all those 30-39 years of age and over
j <- 11
i <- 7:8
p[i, j] <- uptake_by_age[i] * (1 - prop_hcw_by_age[i]) * (1 - prop_very_vulnerable_by_age[i] - prop_underlying_condition_by_age[i])
## JCVI group 11: all those 18-29 years of age and over
j <- 12
i <- 1:6
p[i, j] <- uptake_by_age[i] * (1 - prop_hcw_by_age[i]) * (1 - prop_very_vulnerable_by_age[i] - prop_underlying_condition_by_age[i])


## check 
all(signif(rowSums(p), 5) == signif(uptake_by_age, 5))

################################################################################
## calculate number to vaccinate by age and priority group: n
################################################################################

pop_by_age <- sircovid:::carehomes_parameters(1, region)$N_tot
pop_by_age_mat <- matrix(rep(pop_by_age, n_priority_groups), nrow = n_age_groups)
n <- p * pop_by_age_mat

## check 
all(signif(rowSums(n), 5) == signif(uptake_by_age * pop_by_age, 5))

## plot
par(las = 1, mfrow = c(1, 2))
# hcl.pals()
image(x = 1:n_priority_groups, y = 1:n_age_groups,
      z = t(p),
      xlab = "Priority group",
      ylab = "Age group",
      col = c("white", hcl.colors(10, "Blues", rev = TRUE)),
      breaks = c(0, 1e-18, seq(0.1, 1, 0.1)),
      main = "Proportion to vaccinate")
abline(v = (1:(n_priority_groups + 1)) - 0.5, col = "lightgrey")
abline(h = (1:(n_age_groups + 1)) - 0.5, col = "lightgrey")

image(x = 1:n_priority_groups, y = 1:n_age_groups,
      z = t(n),
      xlab = "Priority group",
      ylab = "Age group",
      col = c("white", hcl.colors(10, "Reds", rev = TRUE)),
      breaks = c(0, 1e-18, seq(60000, 600000, 60000)),
      main = "Number to vaccinate")
abline(v = (1:(n_priority_groups + 1)) - 0.5, col = "lightgrey")
abline(h = (1:(n_age_groups + 1)) - 0.5, col = "lightgrey")

## TODO: add legends
## TODO: could add a column showing overall uptake and overall number to vaccinate by age

################################################################################
## code to generate the input matrix for sircovid
################################################################################
mean_days_between_doses <- 12 * 7
# first doses
first_daily_doses <- 0 * daily_doses
n_to_vaccinate_mat <- array(0, dim = c(n_age_groups, n_priority_groups, length(daily_doses)))
n_to_vaccinate <- array(0, dim = c(n_age_groups, length(daily_doses)))
n_left <- n
# second doses
second_daily_doses <- 0 * daily_doses
n_to_vaccinate_second_dose <- array(0, dim = c(n_age_groups, length(daily_doses)))

for (t in seq_along(daily_doses)) {
  
  ## split doses between first and second doses
  if (t <= mean_days_between_doses) { # only distribute first doses
    second_daily_doses[t] <- 0  
    first_daily_doses[t] <- daily_doses[t]
  } else { # prioritise second doses
    second_daily_doses[t] <- min(daily_doses[t], first_daily_doses[t - mean_days_between_doses])
    first_daily_doses[t] <- daily_doses[t]  - second_daily_doses[t] 
  }
  
  ## allocate first doses
  eligible <- colSums(n_left)
  ## vaccinate fully the top priority groups
  n_full_vacc <- findInterval(first_daily_doses[t], cumsum(eligible)) # vaccinate everyone in these groups
  if (n_full_vacc > 0) {
    n_to_vaccinate_mat[, 1:n_full_vacc, t] <- n_left[,1:n_full_vacc] 
  }
  ## then go down to next priority group
  if (n_full_vacc < n_priority_groups) {
    if(n_full_vacc == 0) {
      remaining_eligible <- first_daily_doses[t]
    } else {
      remaining_eligible <- first_daily_doses[t] - cumsum(eligible)[n_full_vacc]
    }
    ## split remaining doses according to age
    n_to_vaccinate_mat[, n_full_vacc + 1, t] <- round(remaining_eligible * n_left[,n_full_vacc + 1] / sum(n_left[,n_full_vacc + 1] ))
  }
  
  n_left <- n_left - n_to_vaccinate_mat[, , t]
  
  n_to_vaccinate[, t] <- rowSums(n_to_vaccinate_mat[, , t])
  
  ## allocate second doses 12 weeks later
  if (t > mean_days_between_doses) { # also do second doses
    n_to_vaccinate_second_dose[, t] <- n_to_vaccinate[, t - mean_days_between_doses]
  }
  
} # end loop over time

## check this works:
par(mfrow = c(3, 3))
for(t in 1:9) {
  image(t(n_to_vaccinate_mat[, , t]))
}

## summary of number of vaccination in each age group over time
par(mfrow = c(1, 1))
image(t(n_to_vaccinate))

## check the number of doses per day are what we want
plot(colSums(n_to_vaccinate), type = "l", col = "blue", xlab = "Day", ylab = "Daily number of doses")
lines(colSums(n_to_vaccinate_second_dose), type = "l", col = "red")
legend("topright", c("1st doses", "2nd doses"), col = c("blue", "red"), lty = 1)

```

**Modifications needed to sircovid:**
Make sircovid use `n_to_vaccinate` and `n_to_vaccinate_second_dose` above as inputs, where `n_to_vaccinate[i, t]` is the number of people in age group i who should receive a first dose of vaccine at time t (and same for `n_to_vaccinate_second_dose` and second doses).
Inside sircovid, at time t, we should count the "candidates" for first dose vaccination, i.e.: `candidates[i] = S + E + I(non symptomatic) + R in age group i and non vaccinated`. For each age group i, draw `n_vaccinated[i] <- min(candidates[I], n_to_vaccinate[i, t])` 1st dose vaccinations and split those between the relevant S, E, I non symptomatic and R in proportion to the numbers in each of those.
Then similarly  count the "candidates" for second dose vaccination, i.e.: `candidates_2[i] = S + E + I(non symptomatic) + R in age group i and having received 1 dose of vaccine`. For each age group i, draw `n_vaccinated_2[i] <- min(candidates_2[I], n_to_vaccinate_second_dose[i, t])` 2nd dose vaccinations and split those between the relevant S, E, I non symptomatic and R in proportion to the numbers in each of those.

**Notes:**
Need to translate the above which I wrote thinking in days into timesteps.

### PR DESCRIPTION
**Pull request by richfitz**
Move to precomputed vaccine schedule

This is an @annecori/@richfitz collab, and seems to work - changes added to both spim fits now

Adds support designed to use with the in-development projections task

Fixes #226 
Closes #218

### CODE REVIEW

File: R/vaccination.R [hunk 1]
```diff
@@ -121,8 +122,555 @@ build_vaccine_progression_rate <- function(vaccine_progression_rate,
...
+    uptake <- rep(1, n_groups)
+  } else if (length(uptake) == 1L) {
+    uptake <- rep(uptake, n_groups)
+  } else if (length(uptake) != n_groups) {
+    stop(sprintf("Invalid length %d for 'uptake', must be 1 or %d",
+                 length(uptake), n_groups))
+  }
+
+  prop_hcw <- prop_hcw %||%
+    c(rep(0, 4), rep(0.1, 10), rep(0, 5))
```
Comment (edknock) [inline R/vaccination.R:209]: Might be useful to comment here what this represents i.e. 10% of ages 20-69

File: R/vaccination.R [hunk 2]
```diff
@@ -121,8 +122,555 @@ build_vaccine_progression_rate <- function(vaccine_progression_rate,
...
+  prop_very_vulnerable <- prop_very_vulnerable %||%
+    c(rep(0, 4), rep(0.05, 5), rep(0.1, 5), rep(0.15, 5))
+  prop_underlying_condition <- prop_underlying_condition %||%
+    c(rep(0, 4), rep(0.05, 5), rep(0.1, 5), rep(0.15, 5))
```
Comment (edknock) [inline R/vaccination.R:213]: We have very vulnerable and underlying condition proportions for CHR and CHW here - is this necessary given that they are all in the first priority group anyway?

File: R/vaccination.R (hunk 1 above)
Comment (edknock) [reply R/vaccination.R:209]: (And the same below)

File: R/vaccination.R (hunk 2 above)
Comment (annecori) [reply R/vaccination.R:213]: they should not be in the same priority group - can you check they are indeed not @edknock ?

Comment (annecori) [reply R/vaccination.R:213]: one is in priority group 4, the other in group 6 
image

Comment (annecori) [reply R/vaccination.R:213]: ah sorry I misread your message - I see your question is just for CHR and CHW. Indeed it's not useful but I just think it's easier to manipulate vectors of length 19 all the way through. The values I put for now are totally rubbish anyway, and any proportion > 1 in CHR and CHW should not be used anyway

Comment (edknock) [reply R/vaccination.R:213]: Okay, sounds fine to me then

**Review by edknock — APPROVED** [review]

**Pull request merged**